### PR TITLE
Implement tests

### PR DESCRIPTION
**Pull request by yuxqiu**
- [x] `truelearn.models`
- [x] `truelearn.preprocessing`
- [x] `truelearn.datasets`
- [x] `truelearn.learning`
- [x] Improve test coverage

### CODE REVIEW

File: truelearn/learning/_base.py [hunk 1]
```diff
@@ -24,12 +24,17 @@ class BaseClassifier(ABC):
...
-    as it's possible for a type to accept more than one type.
-    To do the constraint check based on this, simply call `self._validate_params`
+    or a tuple of values as it's possible for a type to accept more than one type/value.
+    To do the constraint check based on this, simply call `self._validate_params()`
     in your classifier.
     """
 
     __DEEP_PARAM_DELIMITER: Final[str] = "__"
+
+    # TODO: use constraint and satisfies to validate parameters
```
Comment (KD-7) [inline truelearn/learning/_base.py:34]: I see there is a TODO here

File: truelearn/utils/metrics/_base.py [hunk 2]
```diff
@@ -1,59 +1,7 @@
...
-
-    Args:
-        pred_labels: An iterable of predicted labels
-        act_labels: An iterable of actual labels
-
-    Returns:
-        The f1 score.
-    """
-    return float(metrics.f1_score(pred_labels, act_labels))
+get_precision_score = metrics.precision_score
```
Comment (KD-7) [inline truelearn/utils/metrics/_base.py:4]: Will we still be able to get the docstrings for this?

Comment (yuxqiu) [reply truelearn/utils/metrics/_base.py:4]: It works, but it comes with some warnings.

Comment (yuxqiu) [reply truelearn/utils/metrics/_base.py:4]: It looks like this:

<details>
<summary>picture here</summary>

image

</details>

Comment (yuxqiu) [reply truelearn/utils/metrics/_base.py:4]: I guess it's because the docstring references something in scikit learn module

```
TrueLearn-python-library/truelearn/utils/metrics/__init__.py:docstring of sklearn.metrics._classification.accuracy_score:1: WARNING: undefined label: 'accuracy_score'
TrueLearn-python-library/truelearn/utils/metrics/__init__.py:docstring of sklearn.metrics._classification.accuracy_score:5: WARNING: undefined label: 'precision_recall_f_measure_metrics'
TrueLearn-python-library/truelearn/utils/metrics/__init__.py:docstring of sklearn.metrics._classification.accuracy_score:1: WARNING: undefined label: 'precision_recall_f_measure_metrics'
TrueLearn-python-library/truelearn/utils/metrics/__init__.py:docstring of sklearn.metrics._classification.accuracy_score:1: WARNING: undefined label: 'precision_recall_f_measure_metrics'
```

Comment (yuxqiu) [reply truelearn/utils/metrics/_base.py:4]: I will restore it to its original form.

File: truelearn/learning/_base.py (hunk 1 above)
Comment (yuxqiu) [reply truelearn/learning/_base.py:34]: I'll do this later in another branch.

File: truelearn/utils/metrics/_base.py (hunk 2 above)
Comment (KD-7) [reply truelearn/utils/metrics/_base.py]: > I will restore it to its original form.

This is not with the new way of generating the docs though, I'll check how it works with #31.

**Comment by KD-7** [comment]
Quick note, if you keep deleting codecov messages it will repost it (with new commits), but if you leave it, codecov will only update its comment not make a new one.

**Comment by KD-7** [comment]
I think we can easily bump up coverage to 100% by adding a few more test cases.
If you look here:
https://app.codecov.io/gh/comp0016-group1/truelearn/tree/impl-tests/
It has detailed info about what lines are missed and other info for constructing additional tests.

If I have time, I will try to add some more today.

**Comment by yuxqiu** [comment]
The reason I didn't test for invalid API keys is that we ultimately want users to be able to run these tests in their machines. So even if the API doesn't exist, they should still be able to pass the test. Therefore, instead of letting the tests fail, we should skip the tests and provide some reasons.

**Comment by yuxqiu** [comment]
> 

Sure.

**Comment by KD-7** [comment]
> The reason I didn't test for invalid API keys is that we ultimately want users to be able to run these tests in their machines. So even if the API doesn't exist, they should still be able to pass the test. Therefore, instead of letting the tests fail, we should skip the tests and provide some reasons.

Thats true, but with wikifier it takes a minute to get an API key. Given that only developers are running the tests, we could easily state in the docs to set an environment variable before running the tests with the API key for wikifier. This prevents the skipped tests being hidden when they should flag an issue.

**Comment by yuxqiu** [comment]
> > The reason I didn't test for invalid API keys is that we ultimately want users to be able to run these tests in their machines. So even if the API doesn't exist, they should still be able to pass the test. Therefore, instead of letting the tests fail, we should skip the tests and provide some reasons.
> 
> Thats true, but with wikifier it takes a minute to get an API key. Given that only developers are running the tests, we could easily state in the docs to set an environment variable before running the tests with the API key for wikifier. This prevents the skipped tests being hidden when they should flag an issue.

I think it's a good idea to state this in the documentation. But I don't think the skipped tests are hidden as pytest clearly shows that some tests are skipped and offers some reasons.

Then, if developers see these tests being skipped but want to take advantage of them (perhaps they are experimenting with wikifier code), they can easily register an api key by following our instructions in the documentation and tests. If they don't care about skipped tests (maybe they only care about our learning subpackage), they don't need to worry about the skipped tests.

**Comment by yuxqiu** [comment]
@KD-7 I have completed this PR.

**Comment by KD-7** [comment]
I see you changed the CI @yuxqiu from testing different python versions to different OS

File: truelearn/learning/_base.py [hunk 3]
```diff
@@ -148,92 +153,77 @@ def set_params(self, **args) -> Self:
...
+        self._validate_params()
+
         return self
 
     @final
-    def _validate_params(self, **kargs) -> None:
+    def _validate_params(self) -> None:
         """Validate types of given arguments in __init__.
 
         Args:
```
Comment (KD-7) [inline truelearn/learning/_base.py:178]: Remove this as there is no kargs

File: truelearn/models/_learner_model.py [hunk 4]
```diff
@@ -66,28 +66,25 @@ class LearnerMetaModel:
...
-    novelty_weight: Dict[str, float] = dataclasses.field(
-        default_factory=lambda: {"mean": 0.0, "variance": 1.0}
-    )
-    interest_weight: Dict[str, float] = dataclasses.field(
-        default_factory=lambda: {"mean": 0.0, "variance": 1.0}
-    )
-    bias_weight: Dict[str, float] = dataclasses.field(
-        default_factory=lambda: {"mean": 0.0, "variance": 1.0}
-    )
+    novelty_weight: Weights = Weights()
```
Comment (KD-7) [inline truelearn/models/_learner_model.py:88]: Loads of parameter mean unfilled and variance unfilled errors for 88 to 90

File: truelearn/tests/test_datasets.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        enable_socket()
+
+    def test_load_peek_dataset_with_limit(self):
+        train, test, mapping = datasets.load_peek_dataset(train_limit=1, test_limit=1)
+
+        assert len(train) == 1
+        assert len(test) == 1
+        assert len(mapping) == 30367
+
+    def test_load_peek_dataset_with_invalid_limit(self):
```
Comment (KD-7) [inline truelearn/tests/test_datasets.py:100]: Code duplication for this whole test with line 155, I think its a mistake because the test is indentical you arent using peek dataset raw in 155

**Review by KD-7 — COMMENTED** [review]
I am getting a lot of warnings about unfilled parameters when you use Weights()

File: truelearn/models/_learner_model.py (hunk 4 above)
Comment (yuxqiu) [reply truelearn/models/_learner_model.py:88]: This is weird. I don't have this warning. We could use `Weights' in this way, since we have defined `defaults'.

File: truelearn/learning/_base.py (hunk 3 above)
Comment (yuxqiu) [reply truelearn/learning/_base.py:178]: There is no `kargs` in the latest version. Looks like you are reading a previous version?

**Comment by yuxqiu** [comment]
prospector issues a warning ("multiple statements on one line (def)") after you change the `PEEKKnowledgeComponentGenerator`. I guess we need to put `...` in the new line.

**Comment by KD-7** [comment]
> prospector issues a warning ("multiple statements on one line (def)") after you change the `PEEKKnowledgeComponentGenerator`. I guess we need to put `...` in the new line.

About that, I dont think we test the __call__ method, its the only thing thats showing up as missed coverage

**Comment by yuxqiu** [comment]
> > prospector issues a warning ("multiple statements on one line (def)") after you change the `PEEKKnowledgeComponentGenerator`. I guess we need to put `...` in the new line.
> 
> About that, I dont think we test the **call** method, its the only thing thats showing up as missed coverage

That's true. But unfortunately, we cannot test it as we cannot instantiate a class that inherits `Protocol`.

File: truelearn/models/_learner_model.py (hunk 4 above)
Comment (KD-7) [reply truelearn/models/_learner_model.py:88]: Yea turns out to be something Pycharm doesnt want to fix:
https://youtrack.jetbrains.com/issue/PY-40405

There is a workaround, but im wondering if its even worth it

Comment (yuxqiu) [reply truelearn/models/_learner_model.py:88]: As we are using `typing`, let's fix it.

**Comment by KD-7** [comment]
> > I see you changed the CI @yuxqiu from testing different python versions to different OS
> 
> Yes. I thought it would take up our free time, but I just confirmed it doesn't. So let's put it back on?



> > I see you changed the CI @yuxqiu from testing different python versions to different OS
> 
> Yes. I thought it would take up our free time, but I just confirmed it doesn't. So let's put it back on?



> > I see you changed the CI @yuxqiu from testing different python versions to different OS
> 
> Yes. I thought it would take up our free time, but I just confirmed it doesn't. So let's put it back on?



> > I see you changed the CI @yuxqiu from testing different python versions to different OS
> 
> Yes. I thought it would take up our free time, but I just confirmed it doesn't. So let's put it back on?

Yep, I think we can also keep the testing on different OS's as well as different python versions

File: truelearn/learning/_base.py (hunk 3 above)
Comment (KD-7) [reply truelearn/learning/_base.py:178]: Sorry,im referring to the docstring, args can be removed completely

File: truelearn/tests/test_preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            def __init__(self, *_args, **_kargs):
+                ...
+
+            def __enter__(self):
+                return self
+
+            def __exit__(self, *_args, **_kargs):
+                ...
+
+            def read(self):
```
Comment (KD-7) [inline truelearn/tests/test_preprocessing.py:129]: read may be static

**Review by KD-7 — COMMENTED** [review]
I am getting a lot of warnings about unfilled parameters when you use Weights()

**Comment by KD-7** [comment]
Im getting quite a lot of complaints of code duplication in the unit tests, Im wondering if we can just leave it as is
Im referring to:
image

**Comment by yuxqiu** [comment]
> Im getting quite a lot of complaints of code duplication in the unit tests, Im wondering if we can just leave it as is

Yep. Extracting these "duplicate" code into some shared functions will only reduce the readability of the code.

**Comment by KD-7** [comment]
I will approve once you remove Args in the docstring

**Comment by KD-7** [comment]
> I will approve once you remove Args in the docstring (see my review)

**Comment by yuxqiu** [comment]
> I will approve once you remove Args in the docstring (see my review)

Done.

**Review by KD-7 — APPROVED** [review]
Ready for 🚀

**Comment by yuxqiu** [comment]
> Ready for 🚀

Let's go 🚀
